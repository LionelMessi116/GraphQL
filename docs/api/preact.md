---
title: @urql/preact
order: 2
---

# @urql/preact

Please refer to [the React API section](/api/urql) for details of the Preact API.
