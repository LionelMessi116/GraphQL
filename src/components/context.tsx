import createReactContext from 'create-react-context';
import { Client } from '../types';

const context = createReactContext<Client>({} as any); // Requires default value (which is unused)

export const ContextProvider = context.Provider;
export const ContextConsumer = context.Consumer;
