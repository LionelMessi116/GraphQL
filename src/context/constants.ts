export const CLIENT = '$$_URQL';
