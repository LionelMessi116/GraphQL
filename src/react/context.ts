import createContext from 'create-react-context';
import { Client, createClient } from '../client';

// We assume some default options here; mainly not to actually be used
// but not to error catastrophically if someone is just playing around
const defaultOptions = { url: '/graphql' };

export const Context = createContext<Client>(createClient(defaultOptions));
export const Provider = Context.Provider;
export const Consumer = Context.Consumer;
