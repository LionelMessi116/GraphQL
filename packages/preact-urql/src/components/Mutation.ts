import { VNode } from 'preact';
import { DocumentNode } from 'graphql';
import { OperationResult, OperationContext } from '@urql/core';
import { useMutation, UseMutationState } from '../hooks';

export interface MutationProps<T, V> {
  query: DocumentNode | string;
  children: (arg: MutationState<T, V>) => VNode<any>;
}

export interface MutationState<T, V> extends UseMutationState<T> {
  executeMutation: (
    variables?: V,
    context?: Partial<OperationContext>
  ) => Promise<OperationResult<T>>;
}

export function Mutation<T = any, V = any>(
  props: MutationProps<T, V>
): VNode<any> {
  const mutationState = useMutation<T, V>(props.query);
  return props.children({
    ...mutationState[0],
    executeMutation: mutationState[1],
  });
}
