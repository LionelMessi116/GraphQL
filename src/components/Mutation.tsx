import React, { Component, FC, ReactNode } from 'react';
import { pipe, toPromise } from 'wonka';
import { Client } from '../client';
import { Consumer } from '../context';
import { GraphQLRequest, OperationResult } from '../types';
import { CombinedError, createMutation } from '../utils';

interface MutationHandlerProps {
  client: Client;
  query: GraphQLRequest['query'];
  children: (arg: MutationChildProps) => ReactNode;
}

interface MutationHandlerState {
  fetching: boolean;
  data?: any;
  error?: CombinedError;
}

interface MutationChildProps extends MutationHandlerState {
  executeMutation: (variables?: object) => Promise<OperationResult>;
}

class MutationHandler extends Component<
  MutationHandlerProps,
  MutationHandlerState
> {
  state = {
    fetching: false,
    data: undefined,
    error: undefined,
  };

  render() {
    return this.props.children({
      ...this.state,
      executeMutation: this.executeMutation,
    });
  }

  executeMutation = (variables?: object): Promise<OperationResult> => {
    this.setState({
      fetching: true,
      error: undefined,
      data: undefined,
    });

    const mutation = createMutation(this.props.query, variables);

    return pipe(
      this.props.client.executeMutation(mutation),
      toPromise
    )
      .then(result => {
        const { data, error } = result;
        this.setState({ fetching: false, data, error });
        return result;
      })
      .catch(networkError => {
        const error = new CombinedError({ networkError });
        this.setState({ fetching: false, data: undefined, error });
        return { data: undefined, error } as OperationResult;
      });
  };
}

type MutationProps = Exclude<MutationHandlerProps, 'client'>;

export const Mutation: FC<MutationProps> = props => (
  <Consumer>
    {client => <MutationHandler {...props} client={client} />}
  </Consumer>
);
