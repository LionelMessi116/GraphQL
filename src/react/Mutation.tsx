import React, { Component, FC, ReactNode } from 'react';
import { pipe, toPromise } from 'wonka';
import { Client } from '../client';
import { Consumer } from '../context';
import { CombinedError, createMutation } from '../lib';

interface MutationHandlerProps {
  client: Client;
  query: string;
  children: (arg: MutationChildProps) => ReactNode;
}

interface MutationHandlerState {
  fetching: boolean;
  data?: any;
  error?: CombinedError;
}

interface MutationChildProps extends MutationHandlerState {
  executeMutation: (variables: object) => void;
}

class MutationHandler extends Component<
  MutationHandlerProps,
  MutationHandlerState
> {
  public state = {
    fetching: false,
  };

  public render() {
    return this.props.children({
      ...this.state,
      executeMutation: this.executeMutation,
    });
  }

  private executeMutation = async (variables: object) => {
    if (this.props.query === undefined) {
      return;
    }

    this.setState({
      fetching: true,
      error: undefined,
      data: undefined,
    });

    try {
      const { data, error } = await pipe(
        this.props.client.executeMutation(
          createMutation(this.props.query, variables)
        ),
        toPromise
      );

      this.setState({ fetching: false, data, error });
    } catch (error) {
      this.setState({
        ...this.state,
        fetching: false,
        error,
      });
    }
  };
}

type MutationProps = Exclude<MutationHandlerProps, 'client'>;

export const Mutation: FC<MutationProps> = props => (
  <Consumer>
    {client => <MutationHandler {...props} client={client} />}
  </Consumer>
);
