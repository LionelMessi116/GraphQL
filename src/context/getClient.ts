import { getContext } from 'svelte';
import { Client } from 'urql/core';
import { CLIENT } from '../constants';

export const getClient = (): Client => getContext(CLIENT);
