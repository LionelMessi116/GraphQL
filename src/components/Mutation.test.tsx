import { h } from 'preact';
import { act, cleanup, render } from '@testing-library/preact';
import { pipe, fromValue, delay } from 'wonka';
import { Provider } from '../context';
import { Mutation } from './Mutation';

const mock = {
  executeMutation: jest.fn(() =>
    pipe(fromValue({ data: 1, error: 2, extensions: { i: 1 } }), delay(200))
  ),
};
const client = mock as { executeMutation: jest.Mock };
const query = 'mutation Example { example }';

describe('Mutation', () => {
  beforeEach(() => {
    // eslint-disable-next-line no-console
    console.log(
      'supressing console.error output due to react-test-renderer spam (hooks related)'
    );
    jest.spyOn(global.console, 'error').mockImplementation();
  });

  afterEach(() => {
    cleanup();
  });

  it('Should execute the mutation', done => {
    // eslint-disable-next-line
    let execute = () => {},
      props = {};
    const Test = () => <p>Hi</p>;
    const App = () => {
      // @ts-ignore
      return h(Provider, {
        value: client,
        children: [
          h(
            // @ts-ignore
            Mutation,
            { query },
            ({ data, fetching, error, executeMutation }) => {
              execute = executeMutation;
              props = { data, fetching, error };
              return h(Test, {});
            }
          ),
        ],
      });
    };
    render(h(App, {}));
    expect(client.executeMutation).toBeCalledTimes(0);
    expect(props).toStrictEqual({
      data: undefined,
      fetching: false,
      error: undefined,
    });
    act(() => {
      execute();
    });
    expect(props).toStrictEqual({
      data: undefined,
      fetching: true,
      error: undefined,
    });
    setTimeout(() => {
      expect(props).toStrictEqual({ data: 1, fetching: false, error: 2 });
      done();
    }, 400);
  });
});
