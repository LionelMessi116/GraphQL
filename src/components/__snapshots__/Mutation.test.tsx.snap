// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`on init default values match snapshot 1`] = `
Object {
  "executeMutation": [Function],
  "fetching": false,
}
`;
